presto-main/src/test/java/io/prestosql/sql/query/TestJoin.java

```diff
@@ -50,4 +50,40 @@ public void testCrossJoinEliminationWithOuterJoin()
                         "JOIN d ON d.id = a.id",
                 "VALUES 1");
     }
+
+    @Test
+    public void testInPredicateInJoinCriteria()
+    {
+        assertions.assertQuery(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) JOIN (VALUES 1, 3, NULL) u(x) ON t.x IN (VALUES 1)",
+                "VALUES (1, 1), (1, 3), (1, NULL)");
+
+        assertions.assertQuery(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) LEFT JOIN (VALUES 1, 3, NULL) u(x) ON t.x IN (VALUES 1)",
+                "VALUES (1, 1), (1, 3), (1, NULL), (2, NULL), (NULL, NULL)");
+
+        assertions.assertQuery(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) RIGHT JOIN (VALUES 1, 3, NULL) u(x) ON t.x IN (VALUES 1)",
+                "VALUES (1, 1), (1, 3), (1, NULL)");
+
+        assertions.assertQuery(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) FULL JOIN (VALUES 1, 3, NULL) u(x) ON t.x IN (VALUES 1)",
+                "VALUES (1, 1), (1, 3), (1, NULL), (2, NULL), (NULL, NULL)");
+
+        assertions.assertQuery(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) JOIN (VALUES 1, 3, NULL) u(x) ON u.x IN (VALUES 1)",
+                "VALUES (1, 1), (2, 1), (NULL, 1)");
+
+        assertions.assertQuery(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) LEFT JOIN (VALUES 1, 3, NULL) u(x) ON u.x IN (VALUES 1)",
+                "VALUES (1, 1), (2, 1), (NULL, 1)");
+
+        assertions.assertQuery(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) RIGHT JOIN (VALUES 1, 3, NULL) u(x) ON u.x IN (VALUES 1)",
+                "VALUES (1, 1), (2, 1), (NULL, 1), (NULL, 3), (NULL, NULL)");
+
+        assertions.assertQuery(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) FULL JOIN (VALUES 1, 3, NULL) u(x) ON u.x IN (VALUES 1)",
+                "VALUES (1, 1), (2, 1), (NULL, 1), (NULL, 3), (NULL, NULL)");
+    }
```

Comment:
Please add more tests to show the scope of this change:
- not only in-predicates, but also quantified comparisons benefit from being planned on top of the appropriate source plan:
`SELECT * FROM (VALUES 1, 2, NULL) t(x) RIGHT JOIN (VALUES 1, 3, NULL) u(x) ON u.x > ALL (VALUES 1)` failed but now it works
- outer join + mixed left and right references still fails:
`SELECT * FROM (VALUES 1, 2, NULL) t(x) RIGHT JOIN (VALUES 1, 3, NULL) u(x) ON t.x + u.x > ALL (VALUES 1)`
- inner join is unaffected by this change and works with mixed references:
`SELECT * FROM (VALUES 1, 2, NULL) t(x) JOIN (VALUES 1, 3, NULL) u(x) ON t.x + u.x > ALL (VALUES 1)`